sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1286,30 +1370,35 @@ class WorkerHandlerManager(object):
   def __init__(self, environments, job_provision_info=None):
     self._environments = environments
     self._job_provision_info = job_provision_info
-    self._cached_handlers = {}
+    self._cached_handlers = collections.defaultdict(list)
     self._state = FnApiRunner.StateServicer() # rename?
+    # TODO(hannahjiang): create a grpc_server only for grpc env.
+    self.grpc_server = GrpcServer(self._state, self._job_provision_info)
 
-  def get_worker_handler(self, environment_id):
+  def get_worker_handlers(self, environment_id, num_workers):
     if environment_id is None:
       # Any environment will do, pick one arbitrarily.
       environment_id = next(iter(self._environments.keys()))
     environment = self._environments[environment_id]
-
-    worker_handler = self._cached_handlers.get(environment_id)
-    if worker_handler is None:
-      worker_handler = self._cached_handlers[
-          environment_id] = WorkerHandler.create(
-              environment, self._state, self._job_provision_info)
-      worker_handler.start_worker()
-    return worker_handler
+    worker_handler_list = self._cached_handlers[environment_id]
+    if not worker_handler_list or len(worker_handler_list) < num_workers:
```

Comment:
`not worker_handler_list` will be subsumed by `len(worker_handler_list) < num_workers`.